plugins/amd-gpu/fu-amd-gpu-device.c

```diff
@@ -1,45 +1,275 @@
+		fu_device_set_name(device, (const char *)vbios_info.name);
+
+	return TRUE;
+}
+
+static FuFirmware *
+fu_amd_gpu_device_prepare_firmware(FuDevice *device,
+				   GBytes *fw,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuAmdGpuDevice *self = FU_AMDGPU_DEVICE(device);
+	g_autoptr(FuFirmware) firmware = fu_amd_gpu_psp_firmware_new();
+	g_autoptr(FuFirmware) ish_a = NULL;
+	g_autoptr(FuFirmware) partition_a = NULL;
+	g_autoptr(FuFirmware) csm = NULL;
+	const gchar *fw_pn;
+
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return NULL;
+
+	/* we will always flash the contents of partition A */
+	ish_a = fu_firmware_get_image_by_id(firmware, "ISH_A", error);
+	if (ish_a == NULL)
+		return NULL;
+	partition_a = fu_firmware_get_image_by_id(ish_a, "PARTITION_A", error);
+	if (partition_a == NULL)
+		return NULL;
+	csm = fu_firmware_get_image_by_id(partition_a, "ATOM_CSM_A", error);
+	if (csm == NULL)
+		return NULL;
+
+	fw_pn = fu_amd_gpu_atom_get_vbios_pn(csm);
+	if (g_strcmp0(fw_pn, self->vbios_pn) != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "firmware for %s does not match %s",
+			    fw_pn,
+			    self->vbios_pn);
+		return NULL;
+	}
+
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_amd_gpu_device_write_firmware(FuDevice *device,
+				 FuFirmware *firmware,
+				 FuProgress *progress,
+				 FwupdInstallFlags flags,
+				 GError **error)
+{
+	g_autofree gchar *psp_vbflash_status = NULL;
+	g_autofree gchar *psp_vbflash = NULL;
+	g_autoptr(FuIOChannel) image_io = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	const gchar *base;
+	gint retries;
+
+	base = fu_udev_device_get_sysfs_path(FU_UDEV_DEVICE(device));
+	psp_vbflash = g_build_filename(base, "psp_vbflash", NULL);
+
+	image_io = fu_io_channel_new_file(psp_vbflash, error);
+	if (image_io == NULL)
+		return FALSE;
+
+	fu_progress_set_id(progress, G_STRLOC);
+
+	/* stage the image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	if (!fu_io_channel_write_bytes(image_io, fw, 100, FU_IO_CHANNEL_FLAG_NONE, error))
+		return FALSE;
+
+	/* trigger the update (this looks funny but amdgpu returns 0 bytes) */
+	if (!fu_io_channel_read_raw(image_io, NULL, 0, NULL, 100, FU_IO_CHANNEL_FLAG_NONE, NULL))
+		g_debug("triggered update");
+
+	/* poll for completion */
+	psp_vbflash_status = g_build_filename(base, "psp_vbflash_status", NULL);
+	for (retries = 0; retries < PSPVBFLASH_MAX_POLL; retries++) {
+		g_autofree gchar *buf = NULL;
+		gsize sz = 0;
+		guint64 val;
+
+		if (!g_file_get_contents(psp_vbflash_status, &buf, &sz, error))
+			return FALSE;
+		if (!fu_strtoull(buf, &val, 0, G_MAXUINT64, error))
+			return FALSE;
+		if (val == PSPVBFLASH_SUCCESS)
+			break;
+		fu_device_sleep(device, 100);
+	}
+	if (retries == PSPVBFLASH_MAX_POLL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "timeout waiting for pspvbflash to complete");
 		return FALSE;
 	}
```

Comment:
fu_device_retry_full might be more elegant here.

Reply:
@superm1 i've pushed a fixup, can you check that makes sense pls. then LGTM.

Reply:
thanks!